tests/unit/test_cv_device_tools.py

```diff
@@ -0,0 +1,150 @@
+    """
+    @pytest.mark.parametrize(
+        "expected",
+        [
+            (True),    #  success
+            (False),   # failure
+        ],
+    )
+    def test_apply_bundle(self, setup, expected):
+        """
+        Test when current_image_bundle and assigned_image_bundle are different
+
+        :param setup: fixture
+        :param expected: output expected from apply_bundle
+
+        """
+        user_topology = DeviceInventory(data=device_data)
+        _, mock__get_device, cv_tools, mock_get_container_current = setup
+        mock_get_container_current.return_value = current_container_info
+        mock__get_device.return_value = cv_data
+        if not expected:
+            # removing node_id for failure test-case
+            image_bundle['id'] = None
+
+        result = cv_tools.apply_bundle(user_inventory=user_topology)
+        assert result[0].success == expected
+        assert result[0].changed == expected
+
+        # resetting imageBundle id
+        image_bundle['id'] = 'imagebundle_1658329041200536707'
+
+    def test_apply_bundle_check_mode_true(self, setup):
+        """
+        Test when check_mode is true
+        :param setup: fixture
+
+        """
+        user_topology = DeviceInventory(data=device_data)
+        _, mock__get_device, cv_tools, mock_get_container_current = setup
+        mock_get_container_current.return_value = current_container_info
+        mock__get_device.return_value = cv_data
+        cv_tools.check_mode = True
+
+        result = cv_tools.apply_bundle(user_inventory=user_topology)
+        assert result[0].success == True
+        assert result[0].changed == False
+        assert result[0].taskIds == ["check_mode"]
+
+    def test_apply_bundle_same_image(self, setup):
+        """
+        Test when current_image_bundle and assigned_image_bundle are same
+        :param setup: fixture
+
+        if both image bundles are same, nothing to do.
+        """
+        user_topology = DeviceInventory(data=device_data)
+        _, mock__get_device, cv_tools, mock_get_container_current = setup
+        mock_get_container_current.return_value = current_container_info
+
+        cv_data_same = deepcopy(cv_data)
+        cv_data_same['imageBundle']['bundleName'] = device_data[0]['imageBundle']
+        mock__get_device.return_value = cv_data_same
+
+        result = cv_tools.apply_bundle(user_inventory=user_topology)
+        assert result[0].success is False
+        assert result[0].changed is False
+
+    def test_apply_bundle_cvp_api_error(self, setup):
+        """
+        Test for CvpApiError.
+        :param setup: fixture
+
+        """
+        user_topology = DeviceInventory(data=device_data)
+        mock_ansible_module, mock__get_device, cv_tools, mock_get_container_current = setup
+        mock_get_container_current.return_value = current_container_info
+        mock__get_device.return_value = cv_data
+        image_bundle['id'] = 'error_id'
```

Comment:
you could instead raise CvpApiError from a the mocked apply_image_to_element inside the test rather than playing with the function input
```
cv_tools.apply_image_to_element.side_effect = CvpApiError(...)
```

Reply:
apply_image_to_element is not a method of cv_tools

Reply:
yeah cv_tools.api - you get the idea.

Reply:
ya I tried it, cv_tools.__cv_client.api.apply_image_to_element could not be accessed and with the mock object when I tried giving `mock_cvpClient.api.apply_image_to_element.side_effect = CvpApiError(msg='Image bundle ID is not valid')`
getting the error - AttributeError: 'method' object has no attribute 'side_effect'

Reply:
Please suggest if I am doing anything wrong here- pasting the whole method below-
` def test_apply_bundle_cvp_api_error(self, setup, mock_cvpClient):
        """
        Test for CvpApiError.
        :param setup: fixture

        """
        user_topology = DeviceInventory(data=device_data)
        mock_ansible_module, mock__get_device, cv_tools, mock_get_container_current = setup
        mock_get_container_current.return_value = current_container_info
        mock__get_device.return_value = cv_data
        mock_cvpClient.api.apply_image_to_element.side_effect = CvpApiError(msg='Image bundle ID is not valid')

        with pytest.raises(SystemExit) as pytest_error:
            _ = cv_tools.apply_bundle(user_inventory=user_topology)
        assert pytest_error.value.code == 1
        expected_call = [call.fail_json(msg='Error applying bundle to device tp-avd-leaf2: Image bundle ID is not valid')]
        assert mock_ansible_module.mock_calls == expected_call`

Reply:
@Shivani-chourasiya 
You need this line before you can apply `side_effect` to `apply_image_to_element`
```
mock_cvpClient.api.apply_image_to_element = MagicMock()
```

Reply:
Thanks @sugetha24, its working